examples/src/test/java/io/grpc/examples/routeguide/RouteGuideClientTest.java

```diff
@@ -0,0 +1,606 @@
+
+            // let the rpc fail
+            responseObserver.onError(fakeError);
+          }
+        };
+    serviceRegistry.addService(listFeaturesImpl.bindService());
+
+    client.listFeatures(1, 2, 3, 4);
+
+    assertEquals(Rectangle.newBuilder()
+                     .setLo(Point.newBuilder().setLatitude(1).setLongitude(2).build())
+                     .setHi(Point.newBuilder().setLatitude(3).setLongitude(4).build())
+                     .build(),
+                 rectangleDelivered.get());
+    ArgumentCaptor<Throwable> errorCaptor = ArgumentCaptor.forClass(Throwable.class);
+    verify(testHelper)
+        .onMessage(responseFeature1);
+    verify(testHelper)
+        .onRpcError(errorCaptor.capture());
+    assertEquals(fakeError.getStatus(), Status.fromThrowable(errorCaptor.getValue()));
+  }
+
+  /**
+   * Example for testing async client-streaming.
+   */
+  @Test
+  public void testRecordRoute() throws InterruptedException {
+    client.setRandom(noRandomness);
+    Point point1 = Point.newBuilder().setLatitude(1).setLongitude(1).build();
+    Point point2 = Point.newBuilder().setLatitude(2).setLongitude(2).build();
+    Point point3 = Point.newBuilder().setLatitude(3).setLongitude(3).build();
+    Feature requestFeature1 =
+        Feature.newBuilder().setLocation(point1).build();
+    Feature requestFeature2 =
+        Feature.newBuilder().setLocation(point2).build();
+    Feature requestFeature3 =
+        Feature.newBuilder().setLocation(point3).build();
+    final List<Feature> features = Arrays.asList(
+        requestFeature1, requestFeature2, requestFeature3);
+    final List<Point> pointsDelivered = new ArrayList<Point>();
+    final RouteSummary fakeResponse = RouteSummary
+        .newBuilder()
+        .setPointCount(7)
+        .setFeatureCount(8)
+        .setDistance(9)
+        .setElapsedTime(10)
+        .build();
+
+    // implement the fake service
+    RouteGuideImplBase recordRouteImpl =
+        new RouteGuideImplBase() {
+          @Override
+          public StreamObserver<Point> recordRoute(
+              final StreamObserver<RouteSummary> responseObserver) {
+            StreamObserver<Point> requestObserver = new StreamObserver<Point>() {
+              @Override
+              public void onNext(Point value) {
+                pointsDelivered.add(value);
+              }
+
+              @Override
+              public void onError(Throwable t) {
+              }
+
+              @Override
+              public void onCompleted() {
+                responseObserver.onNext(fakeResponse);
+                responseObserver.onCompleted();
+              }
+            };
+
+            return requestObserver;
+          }
+        };
+    serviceRegistry.addService(recordRouteImpl.bindService());
+
+    // send requestFeature1, requestFeature2, requestFeature3, and then requestFeature1 again
+    client.recordRoute(features, 4);
+
+    assertEquals(
```

Comment:
Optional:  You can use assertThat()  with a collection and avoid use of Arrays.asList